File: src/main/java/org/example/datnbbook/model/KhachHang.java
package org.example.datnbbook.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.ColumnDefault;
import org.hibernate.annotations.Nationalized;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Entity
public class KhachHang {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "ma_khach_hang", length = 20, unique = true)
    @ColumnDefault("'KH'+right('00000'+CONVERT([nvarchar](5), NEXT VALUE FOR [dbo].[KHSeq]), 5)")
    private String maKhachHang;

    @Column(name = "ho_ten")
    private String hoTen;

    @Column(name = "email")
    private String email;

    @Column(name = "so_dien_thoai")
    private String soDienThoai;

    @Column(name = "ngay_sinh")
    private LocalDate ngaySinh;

    @Column(name = "gioi_tinh")
    private Integer gioiTinh;

    @Column(name = "trang_thai")
    private Integer trangThai;

    @Column(name = "deleted")
    private Boolean deleted = false;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "created_by")
    private String createdBy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "updated_by")
    private String updatedBy;

    @OneToMany(mappedBy = "khachHang", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<DiaChi> danhSachDiaChi = new ArrayList<>();

//    @OneToMany(mappedBy = "khachHang", cascade = CascadeType.ALL, orphanRemoval = true)
//    @JsonIgnore
//    private List<DiaChi> danhSachDiaChi = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        deleted = false;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public void themDiaChi(DiaChi diaChi) {
        danhSachDiaChi.add(diaChi);
        diaChi.setKhachHang(this);
    }

    public void xoaDiaChi(DiaChi diaChi) {
        danhSachDiaChi.remove(diaChi);
        diaChi.setKhachHang(null);
    }
}
//
//    @OneToMany(mappedBy = "idKhachHang")
//    private Set<PhieuGiamGiaKhachHang> phieuGiamGiaKhachHangs = new LinkedHashSet<>();

File: src/frontend/vue-notus-main/vue-notus-main/vue-notus-main/src/models/KhachHang.js
export default class KhachHang {
    constructor(data = {}) {
        this.id = data.id || null;
        this.hoTen = data.hoTen || '';
        this.email = data.email || '';
        this.soDienThoai = data.soDienThoai || '';
        this.ngaySinh = data.ngaySinh || null;
        this.gioiTinh = data.gioiTinh !== undefined ? data.gioiTinh : true;
        this.trangThai = data.trangThai !== undefined ? data.trangThai : true;
        this.createdAt = data.createdAt || null;
        this.createdBy = data.createdBy || null;
        this.updatedAt = data.updatedAt || null;
        this.updatedBy = data.updatedBy || null;
        this.deleted = data.deleted !== undefined ? data.deleted : false;
        this.danhSachDiaChi = Array.isArray(data.danhSachDiaChi) ? data.danhSachDiaChi.map(diaChi => ({
            id: diaChi.id || null,
            diaChiChiTiet: diaChi.diaChiChiTiet || '',
            tinhThanh: diaChi.tinhThanh || '',
            quanHuyen: diaChi.quanHuyen || '',
            xaPhuong: diaChi.xaPhuong || '',
            macDinh: diaChi.macDinh !== undefined ? diaChi.macDinh : false,
            khachHangId: diaChi.khachHangId || this.id
        })) : [];
        // Only set maKhachHang if it exists in the data (for existing customers)
        if (data.maKhachHang) {
            this.maKhachHang = data.maKhachHang;
        }
    }

    toJSON() {
        const json = {
            id: this.id,
            hoTen: this.hoTen,
            email: this.email,
            soDienThoai: this.soDienThoai,
            ngaySinh: this.ngaySinh,
            gioiTinh: this.gioiTinh,
            trangThai: this.trangThai,
            createdAt: this.createdAt,
            createdBy: this.createdBy,
            updatedAt: this.updatedAt,
            updatedBy: this.updatedBy,
            deleted: this.deleted,
            danhSachDiaChi: this.danhSachDiaChi
        };
        // Only include maKhachHang in JSON if it exists (for existing customers)
        if (this.maKhachHang) {
            json.maKhachHang = this.maKhachHang;
        }
        return json;
    }
} 